lib/puppet/provider/base_dsc/powershell.rb

```diff
@@ -34,15 +37,41 @@ def self.template_path
 
 
   def self.powershell_args
-    ['-NoProfile', '-NonInteractive', '-NoLogo', '-ExecutionPolicy', 'Bypass', '-Command']
+    ps_args = ['-NoProfile', '-NonInteractive', '-Sta', '-NoLogo', '-ExecutionPolicy', 'Bypass', '-Command']
+    ps_args << '-' if !Facter.value(:uses_win32console)
+    ps_args
+  end
+
+  def self.upgrade_message
+    <<-UPGRADE
+The current Puppet version is outdated and uses a library that was necessary at
+the time to support a colored console.
+
+Unfortunately this library creates issues for properly redirecting output from a
+shared PowerShell process.  The shared PowerShell process is used on newer
+versions of Puppet to significantly improve the performance of this module.
+
+It is suggested to upgrade to any x64 version of Puppet (including 3.x), or to a
+Puppet version newer than 3.x.'
+    UPGRADE
+  end
+
+  def ps_manager
+    PuppetX::Dsc::PowerShellManager.instance("#{command(:powershell)} #{self.class.powershell_args.join(' ')}")
   end
 
   def exists?
     version = Facter.value(:powershell_version)
+    uses_win32console = Facter.value(:uses_win32console)
     Puppet.debug "PowerShell Version: #{version}"
     script_content = ps_script_content('test')
     Puppet.debug "\n" + script_content
-    output = powershell(self.class.powershell_args, script_content)
+    if uses_win32console
+      Puppet.warning self.class.upgrade_message
```

Comment:
Yeah, agree that the warning is probably best to emit once per run...

Will change after lunch.
On Jan 12, 2016 12:38 PM, "Rob Reynolds" <notifications@github.com> wrote:

> In lib/puppet/provider/base_dsc/powershell.rb
> <https://github.com/puppetlabs/puppetlabs-dsc/pull/165#discussion_r49509669>
> :
>
> >      Puppet.debug "PowerShell Version: #{version}"
> >      script_content = ps_script_content('test')
> >      Puppet.debug "\n" + script_content
> > -    output = powershell(self.class.powershell_args, script_content)
> > +    if uses_win32console
> > +      Puppet.warning self.class.upgrade_message
>
> No way to turn off the warning? This is going to affect all users of
> Puppet 3.x on x86 - it may start to get somewhat annoying to see the
> message TWICE for every DSC resource.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/puppetlabs-dsc/pull/165/files#r49509669>.
>